workspace/src/localization/localization_py/localization_py/particle_filter.py

```diff
@@ -0,0 +1,180 @@
+            self.particle_distr_dist[i].append(distance)
+            self.particle_distr_head[i].append(self.particles[i][2,0]-obs[2,0])
+            if(len(self.particle_distr_dist[i])>10):
+                self.particle_distr_dist[i].pop(0)
+                self.particle_distr_head[i].pop(0)
+
+    def resample(self):
+        """Resample the particles for the next generation.
+
+        Generates a new generation of particles. Computes a cummulative weight, and then generates num_particles random numbers. Then, whichever cummulative range the random number lies in is the particle that gets propogated. If a particle gets propogated more than once, some randomness is added to the state. 
+
+        """
+
+        cumulative_weights = [sum(self.particle_weights[:i+1]) for i in range(self.num_particles)]
+        sampled_points = []
+        for _ in range(self.num_particles):
+            random_num = random.random()  # Generate a random number between 0 and 1
+
+            # Find the interval where the random number falls
+            for i, cum_weight in enumerate(cumulative_weights):
+                if random_num < cum_weight:
+                    sampled_points.append(i)
+                    break
+        already_seen = []
+        new_particles = []
+        new_pdd = []
+        new_pdh = []
+        new_hpx= []
+        new_hpy = []
+        for j in range(0,self.num_particles):
+            i = sampled_points[j]
+            new_pdd.append(self.particle_distr_dist[i].copy())
+            new_pdh.append(self.particle_distr_head[i].copy())
+            new_hpx.append(self.hpx[i].copy())
+            new_hpy.append(self.hpy[i].copy())
+            if i not in already_seen:
+                already_seen.append(i)
+                new_particles.append(self.particles[i].copy())
+            else:
+                new_particles.append(np.array([[self.particles[i][0,0]+random.gauss(0,0.1)],[self.particles[i][1,0]+random.gauss(0,0.1)],[self.particles[i][2,0]],[self.particles[i][3,0]]]))
+        if(len(already_seen)<50):
+            print("Particles Pruned: "+ str(50-len(already_seen)))
+        self.particles = new_particles
+        self.particle_distr_dist = new_pdd
+        self.particle_distr_head = new_pdh
+        self.hpx = new_hpx
+        self.hpy = new_hpy
+                
+
+
+
+    def assign_weights(self):
+        """Assigns weights to each particle.
+
+        Assigns a weight to each particle based on the wasserstein distance between the current particle measurement probability distribution and the expected measurement probability distribution. 90% of weight is assigned to the position, and 10% weight is assigned to the heading. Normalizes the weights after each weight is assigned.
+
+        """
+        for i in range(0,self.num_particles):
+            self.particle_weights[i] = 0.9/wasserstein_distance(self.particle_distr_dist[i], self.dist_distr)+0.1/wasserstein_distance(self.particle_distr_head[i], self.head_distr)
```

Comment:
will open an issue related to moving parameters for PF into a yaml file.

Reply:
@Huzaifg I thought about this a bit more and I don't think that doing this would be the best idea... these are very specific parameters, that don't rely on any sort of external dynamics / configuration (they are fairly static). Having them be parameters would require having a file path that is passed all the way down from the launch file, which creates a lot of clutter and disorganization for a parameter that will pretty much never get changed... I can change this if you feel strongly about this though

Reply:
If you think that these parameters should not change then no problem. I don't feel strongly about giving control of these parameters to the user